### PR TITLE
Fix Issue 22309 - Taking the address of a stack variable struct with this is wrongly seen as @safe

### PR DESCRIPTION
**Pull request by iK4tsu**
Signed-off-by: João Lourenço <jlourenco5691@gmail.com>

### CODE REVIEW

**Comment by iK4tsu** [comment]
I guess this breaks a lot of existent code. How can I proceed here?

**Comment by RazvanN7** [comment]
One could argue that when dip1000 will become the default this issue will be fixed by default.

**Comment by RazvanN7** [comment]
I wonder why this check was done only for dip1000 and not for normal safe code also.

**Comment by iK4tsu** [comment]
But we shouldn't have to have this safety issue until dip1000 becomes officialized.

**Comment by RazvanN7** [comment]
I was just wondering why this is behind the -dip1000 switch and not in the normal @safe code?

**Comment by iK4tsu** [comment]
I guess this is it https://github.com/dlang/dmd/pull/6172 :thinking:

**Comment by iK4tsu** [comment]
So, the original PR added the fix but only with the `-transition=safe` flag. The reasoning was there was at least a function in Phobos that failed with this change.

> It's been an error for years to take the address of a local in @safe code. It does not make sense to allow taking the address of a field of a local. This was not checked for, and that's what this PR does. It is not part of DIP1000.
>
> It does break at least one Phobos function (the autotester quits at the first error), which is why I amended it to only check if -transition=safe, but the code is incorrect.

This comment by Walter also explicitly says that the change is not a part of `DIP1000` however, the same flag used to fix this was named changed a few times and corresponds to the `useDIP1000` parameter. At the moment some functions in Phobos and even in DMD are relying on this in `@safe` code usage. Should this be dismissed as the first time and now claim that this is indeed part of `DIP1000` or tackle it as a bug and fix the root causes so that this can be merged?

**Comment by RazvanN7** [comment]
I think we can go ahead and fix the existing issues so that this can be merged.

**Comment by RazvanN7** [comment]
@iK4tsu any plans on taking this to the finish line?

**Comment by iK4tsu** [comment]
Yeah, currently I have a lot less time to contribute due to my job. Simple issues I can manage, but this one requires a bit more attention and this definitely went to a lower priority in my list because of that. I still want to finish this, I just don't know when I'm going to have the time to really tackle and spend the time on this issue. I will try to keep this one under my nose and mess with it in my free time over other stuff.

**Comment by iK4tsu** [comment]
Hmm, I took a look at this again and I think that this PR does not fix the issue. I remember at the time trying to get this to work and assuming this was already implemented for dip1000. So what happened was: I found the issue, then I created a local branch to find the cause in the compiler; building and passing the dip1000 preview flag didn't do anything, the problem persisted,  however, removing that bit of code and building the compiler again seemed to work but only in the release binary. Also, it only worked if I didn't pass the dip1000 flag. Basically, I falsely assumed this was the fix because of that quick coincidence. This is strange behavior nonetheless, and if someone has an explanation for it I would appreciate it. But for now, I think we can close this PR because this clearly isn't the right solution.

Since then things improved in dip1000 and *now* I think there are checks for this. For example, the unit test of this PR does not compile with the newest dmd master version. And the error is more in-depth than what I was proposing because it checks for scope parameters. Because the optimal solution involves making use of the scope keyword, which is a thing that just works properly with dip1000, I change my mind about this having to be fixed in non-dip1000 versions.

**Comment by iK4tsu** [comment]
> Your story is a bit hard to follow, because I need to know exactly what you are referring to when you say things like "didn't do anything", "that bit of code", "the release binary" etc.

Ok, so I'll try to explain a bit more descriptively step by step. What I did was:
* performed the change in this PR
* ran `src/build.sh`
* executed the unit test with the generated debug executable: no error
* executed the unit test with the generated release executable: errors with: cannot return this in safe...

However, executing with dip1000 before this change did not trigger the last error in any case. Btw, I was able to reproduce the above even with the latest changes.

**Comment by dkorpel** [comment]
> Btw, I was able to reproduce the above even with the latest changes.

Good catch. Reduced test case:
```D
@safe:
void escape(S* s);
    
struct S
{
    void f() {escape(&this);}
}
```

**Comment by dkorpel** [comment]
It's fixed on the current master though, the nightly on [run.dlang.io](https://run.dlang.io/) is behind.
